src/core/thread/mle_router.cpp

```diff
@@ -447,14 +447,62 @@ void MleRouter::HandleAdvertiseTrickleTimer(void)
 
 void MleRouter::StopAdvertiseTrickleTimer(void) { mAdvertiseTrickleTimer.Stop(); }
 
+#define SPEC_1167
+void MleRouter::RecalculateAdvertiseInterval(void)
+{
+    VerifyOrExit(IsRouterOrLeader());
+
+    {
+        uint32_t neighbors = 0;
+
+        // Router
+        for (uint16_t index = 0; index <= kMaxRouterId; index++)
+        {
+            Router *router = Get<RouterTable>().FindRouterById(static_cast<uint8_t>(index));
+
+            if (router != nullptr && router->IsStateValid())
+            {
+                neighbors++;
+            }
+        }
+
+        uint32_t advertiseIntervalMax = (neighbors + 1) * 4;
+        if (advertiseIntervalMax > kAdvertiseIntervalMax)
+        {
+            advertiseIntervalMax = kAdvertiseIntervalMax;
+        }
+        else if (advertiseIntervalMax < 12)
+        {
+            advertiseIntervalMax = 12;
+        }
```

Comment:
Can use the following macros:
https://github.com/openthread/openthread/blob/874b6beca8c0ce09f2d296652d234c0d011ad1d1/src/core/common/code_utils.hpp#L85-L95